### DISCUSSION TITLE
Auto Updater

### DISCUSSION THREAD

**Discussion by mattahorton** [jacktrip/jacktrip#544] [Ideas]
We should have an auto-updater tool for users that install via installers and application bundles. Users downloading binaries or building on their own don't expect or desire auto-updating. I believe that users installing via package managers get the benefits of auto-updating from those package managers, though if we choose something that can interact with those package managers, that would be great.

Here are a couple options I have found. Please feel free to discuss these, but **also if you have recommendations or concerns I haven't thought of, please provide those**!

I don't know everything about this space, and simply care about helping people stay updated with the latest JackTrip release.

### **QtAutoUpdater** - https://github.com/Skycoder42/QtAutoUpdater

QtAutoUpdater is a tool primarily built to be used on top of the Qt installer framework, but supporting more backends. It would let us configure either a Qt Widgets or QML view for our updater which could pull from these 5 relevant backends (and Google Play, which is irrelevant):

- [**Qt Installer Framework**](https://doc.qt.io/qtinstallerframework/index.html) (For cross platform desktop)
- [**PackageKit**](https://www.freedesktop.org/software/PackageKit/) (Works with most linux package managers)
- [**Chocolatey**](https://chocolatey.org/) (A package manager for Windows)
- [**Homebrew**](https://brew.sh/) (A package manager for macOs)
- **A custom backend** that checks for updates based on a HTTP-request and can download and execute any update tool

Pros:
- This interacts with the Qt installer framework and is built as a Qt module. If our project was setup to work with the Qt installer framework, this would be trivial to add.
- Supports many Linux package managers via PackageKit
- Custom backend could let us use Github releases and bypass the Installer framework entirely. Maybe?

Cons: 
-  We don't use the Qt installer framework to build our installers. That work is very much not trivial. It's not _required_ to use, but it is certainly what the documentation favors and the best option of the 5, followed up by some combo of a custom backend (github releases?) and PackageKit.

### **Sparkle/WinSparkle** - https://sparkle-project.org/ - https://winsparkle.org/

Sparkle is basically the de factor updater on macOS and someone has all but ported it to Windows.

[It can be used with Qt](https://github.com/sparkle-project/Sparkle/issues/1137), and configuration on both macOS and Windows seems messy but less messy than swapping out how we build our installers.

Pros:
- Sparkle is a standard on macOS and there's lots of support for it.
- No need to rewrite our installers.

Cons: 
- Configuration is somewhat complicated for Qt applications, but documented at least.
- Windows port is less well-known.
- No Linux support.

**Comment by mikedickey** [comment]
Auto-updates on Windows and Mac OS X is very much needed!

"No Linux support" is not only fine, but preferable. All Linux distros have their own mechanisms for auto-update already. Having an out-of-band update mechanism would be abnormal and annoying.

**Comment by mattahorton** [comment]
Fwiw, I'd love to hear thoughts from @dyfer @psiborg112 @cchafe @ntonnaett and anyone else on this matter.

Trying to make use of Discussions 😄

**Reply by nwang92** [reply]
+1, I think auto-update is a feature that pays dividends immediately and is also table stakes for a desktop app these days. 

One extra recommendation: do any of those frameworks support the ability to have "Stable" and "Edge" channels? I imagine the more risk-tolerant people would use "Edge" builds from develop, but it would be nice to have the ability to promote known-good builds to "Stable" once they've been battle tested.

**Comment by cchafe** [comment]
I'm also excited about the prospect. Having never implemented an auto-updater I can only imagine the complexities. Very early on our Win10 port was using the Qt installer framework but I doubt going putting effort into that again makes sense. I can vouch that it did work as advertised.

**Comment by mattahorton** [comment]
Here's [another one I found](https://www.dblsqd.com/). It is technically paid, but 1. it's cheap and 2. our project can fit within the free-for-open-source requirements currently.

The work to do with this one is next to nothing, and what is interesting is that I have a hunch their SDK might work with _any_ AppCast feed. If that's the case, we could host our own, use their SDK to implement, and pay nothing.

**Comment by nwang92** [comment]
Another one I found (but admittedly I haven't spent any time looking into how it works) but might as well add it to the list: https://github.com/Mendeley/Update-Installer

**Comment by forsythac** [comment]
Hey guys, we've definitely committed to making auto-update work through dblsqd, which I believe is included in the RC.

What do we do with this discussion? Should it have been converted to an issue or just closed out? Just wondering about github process- right now it looks like this is an open discussion without a conclusion

**Comment by mattahorton** [comment]
@forsythac Let's go with this!

In [this PR](https://github.com/jacktrip/jacktrip/pull/574) (#574) we introduced auto-updating. There is still work being done to make sure this will all run smoothly, but for now let's lock this discussion and we can start a new issue or a new discussion if anything warrants one in the future.
